### PR TITLE
[EVM] Add get random source to Cadence Arch

### PR DESCRIPTION
**Pull request by sideninja**
Closes: #5507 

This PR implements a new function on the Cadence Arch precompile contract, that exposes access to Flow RandomBeaconHistory contract. This allows for retrieving a random source in the EVM which can be used as a seed for any PRG implementations. 

The new method could be called like so:
```solidity
(bool ok, bytes memory data) = cadenceArch.staticcall(abi.encodeWithSignature("getRandomSource(uint64)", height))
```

It also allows for implementing commit-reveal schemes which are required for a lot of applications and was discussed on Discord: https://discord.com/channels/613813861610684416/1230199673629835375
We must keep in mind that the new function for obtaining random source relies on Cadence block heights not EVM block heights, which are not co-related but this is fine since we can use the get latest flow height function also available on Cadence arch. 

@sisyphusSmiling even made a proof of concept that abstracts away this implementation detail https://gist.github.com/sisyphusSmiling/3e58dc3827c69afd038abe65ef8dda92

### CODE REVIEW

File: fvm/evm/testutils/contracts/test.sol [hunk 1]
```diff
@@ -48,6 +48,13 @@ contract Storage {
         selfdestruct(payable(msg.sender));
     }
 
+    function verifyArchCallToRandomSource(uint64 height) public view returns (uint64) {
+        (bool ok, bytes memory data) = cadenceArch.staticcall(abi.encodeWithSignature("getRandomSource(uint64)", height));
+        //require(ok, "unsuccessful call to arch ");
```
Comment (sisyphusSmiling) [inline fvm/evm/testutils/contracts/test.sol:53]: Should this check be commented?

Comment (sideninja) [reply fvm/evm/testutils/contracts/test.sol:53]: nope :D

Comment (sideninja) [reply fvm/evm/testutils/contracts/test.sol:53]: it's not in the binary, just forgot to update the source

File: fvm/evm/evm.go [hunk 2]
```diff
@@ -23,6 +23,10 @@ func StorageAccountAddress(chainID flow.ChainID) (flow.Address, error) {
 	return sc.EVMStorage.Address, nil
 }
 
+func RandomBeaconAddress(chainID flow.ChainID) flow.Address {
+	return systemcontracts.SystemContractsForChain(chainID).RandomBeaconHistory.Address
+}
+
```
Comment (ramtinms) [inline fvm/evm/evm.go:29]: Does this have to be here, given we pass the chainID to the underlying layer. I know we do this for storage account but I think that's a legacy from the time EVM storage address was not part of system contracts.

File: fvm/evm/precompiles/arch_test.go [hunk 3]
```diff
@@ -37,6 +38,34 @@ func TestArchContract(t *testing.T) {
...
 
+	t.Run("test get random source", func(t *testing.T) {
+		address := testutils.RandomAddress(t)
+		rand := uint64(1337)
+		pc := precompiles.ArchContract(
+			address,
+			nil,
+			nil,
+			func(u uint64) (uint64, error) {
+				return rand, nil
```
Comment (ramtinms) [inline fvm/evm/precompiles/arch_test.go:49]: I might add a line here to check if the input that is received is what is expected.

File: fvm/evm/testutils/contracts/test.sol [hunk 4]
```diff
@@ -48,6 +48,13 @@ contract Storage {
         selfdestruct(payable(msg.sender));
     }
 
+    function verifyArchCallToRandomSource(uint64 height) public view returns (uint64) {
+        (bool ok, bytes memory data) = cadenceArch.staticcall(abi.encodeWithSignature("getRandomSource(uint64)", height));
+        require(ok, "unsuccessful call to arch ");
+        uint64 output = abi.decode(data, (uint64));
+        return output;
```
Comment (ramtinms) [inline fvm/evm/testutils/contracts/test.sol:55]: I usually use the verify methods to verify the output matches what is expected, for example see the verifyArchCallToFlowBlockHeight , we check the output produced matches the expected.

File: fvm/evm/precompiles/arch_test.go (hunk 3 above)
Comment (sideninja) [reply fvm/evm/precompiles/arch_test.go:49]: good idea yeah

File: fvm/evm/testutils/contracts/test.sol (hunk 4 above)
Comment (sideninja) [reply fvm/evm/testutils/contracts/test.sol:55]: I did that in Go test, but yeah might be ok to adpot the pattern you made

**Review by janezpodhostnik — APPROVED** [review]
very nice!

**Pull request merged**